File: Scripts/HandleAndVisualizeExpermintalData.py
import numpy as np
import matplotlib.pyplot as plt
from scipy.integrate import cumulative_trapezoid
import h5py
import mplcursors

import os

# --- 1. Load Data ---
# Using the existing .mat file in the directory
script_dir = os.path.dirname(os.path.abspath(__file__))
file_path = os.path.join(script_dir, '3C90_TX-25-15-10_Data1_Cycle.mat')
experiment_id = 1  # 1-based index from MATLAB
idx = experiment_id - 1 # Convert to 0-based index for Python

print(f"Loading {file_path}...")

# Initialize variables to allow scope access later
v_sec = None
i_prim = None
n_prim = None
n_sec = None
ae = None
le = None
dt = None
hdc = None
temperature = None
dutyp = None
frequency = None
flux_cmd = None


with h5py.File(file_path, 'r') as f:
    # Access the Data group
    Data = f['Data']
    
    # Helper to read dataset and handle dimensions
    # MATLAB [rows, cols] -> HDF5 [cols, rows]
    # We need to be careful with column-major vs row-major.
    # usually extracting a column 'idx' in MATLAB (if accessing by row)
    # Wait:
    # MATLAB: V_sec = Data.Voltage(ExperimentID,:); -> Row 'ExperimentID'
    # HDF5 for 'Voltage': Shape will be (N_samples, N_experiments)
    # So we need the 'ExperimentID' column in HDF5.
    
    v_ds = Data['Voltage'] # Shape likely (1000, 60000)
    v_sec = v_ds[:, idx] # Get all samples for this experiment
    
    i_ds = Data['Current']
    i_prim = i_ds[:, idx]
    
    # Scalar/Small arrays
    def get_scalar(key):
        val = np.array(Data[key])
        if val.size == 1:
            return val.item()
        return val

    n_prim = get_scalar('Primary_Turns')
    n_sec = get_scalar('Secondary_Turns')
    ae = get_scalar('Effective_Area')
    le = get_scalar('Effective_Length')
    
    # 1D arrays
    def get_indexed(key, idx):
        val = np.array(Data[key]).flatten()
        if len(val) > idx:
            return val[idx]
        return val.item()

    dt = get_indexed('Sampling_Time', idx)
    hdc = get_indexed('Hdc_command', idx)
    temperature = get_indexed('Temperature_command', idx)
    dutyp = get_indexed('DutyP_command', idx)
    frequency = get_indexed('Frequency_command', idx)
    flux_cmd = get_indexed('Flux_command', idx)

print("Data loaded successfully with h5py.")

# Create time vector
t = np.arange(len(v_sec)) * dt

# --- 2. Calculate Magnetizing Force (H) ---
# H = (N * I) / Le
h = (n_prim * i_prim) / le

# --- 3. Calculate Flux Density (B) ---
# Remove DC offset from voltage
v_clean = v_sec - np.mean(v_sec)

# Integrate Voltage: Integral(V) dt
flux = cumulative_trapezoid(v_sec, t, initial=0)

# B = Flux / (N * Ae)
b = flux / (n_sec * ae)
# Remove DC offset from B
b = b - np.mean(b)

# --- 4. Plotting ---
fig = plt.figure('B-H Analysis', figsize=(12, 8), facecolor='white')

# Subplot 1: Voltage and Current (Top-Left)
ax1 = plt.subplot2grid((2, 2), (0, 0))
color_v = 'tab:blue'
ax1.set_xlabel('Time (s)')
ax1.set_ylabel('Voltage (V)', color=color_v)
ax1.plot(t, v_sec, color=color_v, linewidth=1.5, label='V (Volts)')
ax1.tick_params(axis='y', labelcolor=color_v)
ax1.grid(True)

ax2 = ax1.twinx()
color_i = 'tab:orange'
ax2.set_ylabel('Current (A)', color=color_i)
ax2.plot(t, i_prim, color=color_i, linewidth=1.5, label='I (Amps)')
ax2.tick_params(axis='y', labelcolor=color_i)
ax1.set_title('Voltage & Current')

lines_1, labels_1 = ax1.get_legend_handles_labels()
lines_2, labels_2 = ax2.get_legend_handles_labels()
ax1.legend(lines_1 + lines_2, labels_1 + labels_2, loc='upper right')

# Subplot 2: B and H (Bottom-Left)
ax3 = plt.subplot2grid((2, 2), (1, 0))
color_b = 'tab:blue'
ax3.set_xlabel('Time (s)')
ax3.set_ylabel('Flux Density B (Tesla)', color=color_b)
ax3.plot(t, b, color=color_b, linewidth=1.5, label='B (Tesla)')
ax3.tick_params(axis='y', labelcolor=color_b)
ax3.grid(True)

ax4 = ax3.twinx()
color_h = 'tab:orange'
ax4.set_ylabel('Field Strength H (A/m)', color=color_h)
ax4.plot(t, h, color=color_h, linewidth=1.5, label='H (A/m)')
ax4.tick_params(axis='y', labelcolor=color_h)
ax3.set_title('B & H Waveforms')

lines_3, labels_3 = ax3.get_legend_handles_labels()
lines_4, labels_4 = ax4.get_legend_handles_labels()
ax3.legend(lines_3 + lines_4, labels_3 + labels_4, loc='upper right')

# Subplot 3: B-H Loop (Right side)
ax_bh = plt.subplot2grid((2, 2), (0, 1), rowspan=2)
ax_bh.plot(h, b, linewidth=2)
ax_bh.set_xlabel('Magnetizing Force H (A/m)')
ax_bh.set_ylabel('Flux Density B (Tesla)')
ax_bh.set_title('B-H Hysteresis Loop')
ax_bh.grid(True)
ax_bh.autoscale(enable=True, axis='both', tight=True)

# Add Info Box
info_str = (
    f"Hdc: {hdc} A/m\n"
    f"Temp: {temperature} °C\n"
    f"Freq: {round(frequency/1000)} kHz\n"
    f"Duty: {dutyp}\n"
    f"Flux Cmd: {flux_cmd:.3f} T"
)

props = dict(boxstyle='square', facecolor='white', edgecolor='black', alpha=1.0)
ax_bh.text(0.05, 0.95, info_str, transform=ax_bh.transAxes, verticalalignment='top', bbox=props)

plt.tight_layout()
mplcursors.cursor(hover=True) 
plt.show()

File: Scripts/HandleAndVisualizeExpermintalData.m
%% --- 1. User Inputs (Replace these with your actual values) ---
ExperimentID=1;
V_sec = Data.Voltage(ExperimentID,:);  % Your voltage array (Volts)
I_prim = Data.Current(ExperimentID,:); % Current into Primary (Amps)
N_prim = Data.Primary_Turns;     % Turns on the current-carrying winding (Primary)
N_sec = Data.Secondary_Turns;      % Turns on the voltage-sensing winding (Secondary)
Ae = Data.Effective_Area;            % Cross-sectional area of the core (m^2)
Le = Data.Effective_Length;     % Effective Path Length (meters) -> e.g., 50mm
dt = Data.Sampling_Time(ExperimentID);              % Time step (seconds)
Hdc=Data.Hdc_command(ExperimentID);
Temperature = Data.Temperature_command(ExperimentID);
DutyP = Data.DutyP_command(ExperimentID);
DutyN = Data.DutyN_command(ExperimentID);
Frequency = Data.Frequency_command(ExperimentID);
Flux_cmd = Data.Flux_command(ExperimentID);

Material = Data.Material;
Shape = Data.Shape;
Ve = Data.Effective_Volume;
CoreN = Data.CoreN;
Dataset = Data.Dataset;

% Info Strings
Discarding_info = Data.Discarding_info;
Freq_info = Data.Freq_info;
Cycle_info = Data.Cycle_info;
Date_processing = Data.Date_processing;

% Create time vector (optional, for plotting)
t = (0:length(V_sec)-1) * dt;

% --- 2. Calculate Magnetizing Force (H) ---
% H = (N * I) / Le
H = (N_prim .* I_prim) ./ Le;

% --- 3. Calculate Flux Density (B) ---
% Remove DC offset to prevent integration drift
%V_clean = V_sec - mean(V_sec);

% Integrate Voltage: Integral(V) dt
Flux = cumtrapz(t, V_sec);


% B = Flux / (N * Ae)
B = Flux ./ (N_sec * Ae);
B = B - mean(B);

Energy_Per_Cycle = polyarea(H, B); 

% Power Loss Density (W/m^3) = Energy * Frequency
Pv = Energy_Per_Cycle * Frequency;



% --- 4. Plotting ---
figure('Name', 'B-H Analysis', 'Color', 'white');

% Subplot 1: Time Domain Signals
% Subplot 1: Voltage and Current (Time Domain)
subplot(2,2,1); % Top-Left
yyaxis left
plot(t, V_sec, 'LineWidth', 1.5);
ylabel('Voltage (V)');
xlabel('Time (s)');
ylim auto

yyaxis right
plot(t, I_prim, 'LineWidth', 1.5);
ylabel('Current (A)');
title('Voltage & Current');
legend('V (Volts)', 'I (Amps)');
grid on;

% Subplot 2: B and H (Time Domain)
subplot(2,2,3); % Bottom-Left
yyaxis left
plot(t, B, 'LineWidth', 1.5);
ylabel('Flux Density B (Tesla)');
xlabel('Time (s)');
ylim auto

yyaxis right
plot(t, H, 'LineWidth', 1.5);
ylabel('Field Strength H (A/m)');
title('B & H Waveforms');
legend('B (Tesla)', 'H (A/m)');
grid on;


% Subplot 2: B-H Loop (Hysteresis)
subplot(2,2,[2 4]); % Right side
plot(H, B, 'LineWidth', 2);
xlabel('Magnetizing Force H (A/m)');
ylabel('Flux Density B (Tesla)');
title('B-H Hysteresis Loop');
grid on;
axis tight;

% Add Info Box to Subplot 2 (Top-Left)
InfoString = {
    ['Hdc: ', num2str(Hdc), 'A/m'];
    ['Temp: ', num2str(Temperature), '°C'];
    ['Freq: ', num2str(round(Frequency/1000)), 'kHz'];
    ['Duty: ', num2str(DutyP)];
    ['Flux Cmd: ', num2str(Flux_cmd, '%.3f'), 'T']
    };
text(0.05, 0.95, InfoString, 'Units', 'normalized', 'VerticalAlignment', 'top', ...
    'BackgroundColor', 'white', 'EdgeColor', 'black');

%% % --- 0. Load Data (if not in workspace) ---
% FileName = '3C90_TX-25-15-10_Data1_Cycle.mat';
% if ~exist('Data', 'var')
%     if exist(FileName, 'file')
%         fprintf('Loading %s...\n', FileName);
%         load(FileName);
%     else
%         error('Data structure not found and file %s missing.', FileName);
%     end
% end
% 
% % --- 0.5 Batch Identification: Sinusoidal Excitation ---
% fprintf('Running Sinusoidal Identification on %d experiments...\n', size(Data.Voltage, 1));

% Margin for RMS ~= Peak/sqrt(2)
% Let's use 5% tolerance
SineMargin = 0.005; 

[NumExp, ~] = size(Data.Voltage);
Data.Sinusoidal = zeros(NumExp, 1);

for k = 1:NumExp
    v_sig = Data.Voltage(k, :);
    i_sig = Data.Current(k, :);
    
    % --- FFT Checker Function ---
    % Checks if >90% of AC power is in the fundamental frequency
    check_spectral_purity = @(sig) calculate_purity(sig);
    
    is_v_sine = check_spectral_purity(v_sig);
    is_i_sine = check_spectral_purity(i_sig);
    
    % If either is sinusoidal, mark as 1
    if is_v_sine || is_i_sine
        Data.Sinusoidal(k, 1) = 1;
    else
        Data.Sinusoidal(k, 1) = 0;
    end
end

function is_pure = calculate_purity(sig)
    % Remove DC
    sig_ac = sig - mean(sig);
    L = length(sig_ac);
    
    if max(abs(sig_ac)) < 1e-9
        is_pure = false;
        return;
    end
    
    % FFT
    Y = fft(sig_ac);
    P2 = abs(Y/L);
    P1 = P2(1:floor(L/2)+1);
    P1(2:end-1) = 2*P1(2:end-1);
    PowerSpec = P1.^2;
    
    % Find Fundamental (Max Peak)
    [max_p, ~] = max(PowerSpec);
    total_ac_p = sum(PowerSpec);
    
    if total_ac_p > 0
        purity = max_p / total_ac_p;
        % Threshold: 99% of energy in fundamental
        is_pure = purity > 0.9975; 
    else
        is_pure = false;
    end
end

fprintf('Identification Complete. Sinusoidal Samples: %d / %d\n', sum(Data.Sinusoidal), NumExp);

% Save to new file
[path, name, ext] = fileparts(FileName);
NewName = [name, '_Identified', ext];
save(NewName, 'Data');
fprintf('Saved updated dataset to: %s\n', NewName);
